### ISSUE TITLE
Project ID not being picked up from properties

### ISSUE THREAD

**Issue by joaoandremartins** [spring-cloud/spring-cloud-gcp#375]
I cleared the project ID from my Cloud SDK and I'm specifying it from the properties but, somehow, it isn't in the `GcpProperties` object being autowired by Spring.

**Comment by joaoandremartins** [comment]
This is happening only in certain circumstances, as I find it in one of my toy apps but not in another.

**Comment by joaoandremartins** [comment]
`spring-cloud-gcp-starter-trace` is somehow causing this.

**Comment by joaoandremartins** [comment]
Drilling on this a bit more, it appears that `spring-cloud-starter-sleuth` is causing this..

**Comment by joaoandremartins** [comment]
I get the issue using:

```
        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-web</artifactId>
        </dependency>
        <dependency>
            <groupId>org.springframework.cloud</groupId>
            <artifactId>spring-cloud-gcp-starter</artifactId>
        </dependency>
        <dependency>
            <groupId>org.springframework.cloud</groupId>
            <artifactId>spring-cloud-context</artifactId>
        </dependency>
```

If I remove `spring-cloud-context`, the project ID is correctly processed from the properties.

However, if I manually add every dependency from `spring-cloud-context`, remove `<optional>true</optional>` and remove `spring-cloud-context` itself, the project ID is also successfully discovered.

The best theory I can come up with is that `spring-cloud-context` defines `spring-boot-configuration-processor` as `optional` and because we're not actually passing it anywhere, the properties aren't processed.
But I tried passing it as `<optional>false</optional>` both in `spring-cloud-gcp-autoconfigure` and `spring-cloud-gcp-sample-trace` and the project ID still can't be found.

So I'm a little bit at a miss here... @artembilan would you have any thoughts on this?

**Comment by ChengyuanZhao** [comment]
I'm also stuck on this one, I noticed that in samples where `spring.cloud.gcp.projectId` setting works (like for Storage), the bean function gcpProjectIdProvider isn't called (and a GcpProperties isn't created) until after Tomcat has started, but in this case with sleuth that bean function is needed before Tomcat has started. It seems like `application.properties` isn't available until later on when Tomcat starts?

**Comment by meltsufin** [comment]
I guess one way to fix it is to exclude `spring-cloud-context`.
```xml
        <dependency>
            <groupId>org.springframework.cloud</groupId>
            <artifactId>spring-cloud-gcp-starter-trace</artifactId>
            <exclusions>
                <exclusion>
                    <groupId>org.springframework.cloud</groupId>
                    <artifactId>spring-cloud-context</artifactId>
                </exclusion>
            </exclusions>
        </dependency>
```

**Comment by artembilan** [comment]
Sorry, I don't have enough context for this to determine the issue.

In some place I commented like: let's make a test-case for any external failing use-case!

So, show me, please, the test-case in the project to play with.
Fortunately we have everything we need in our new `autoconfigure` module.

**Comment by meltsufin** [comment]
@ChengyuanZhao Can you try to add a JUnit test that repros this?

**Comment by ChengyuanZhao** [comment]
@meltsufin Do you mean we should add the exclusion in the pom.xml and then add junit test to check that it works? Just two questions

1) should we leave a comment in that xml file like "this is necessary", which would be kind of hacky-feeling and mysterious 
2) i don't think this is possible to produce in unit testing, because getting the GcpProperties bean with the projectId from spring.cloud.gcp.projectId DOES work, it just doesn't work that early in the spring boot web app start up. I can show you what I mean offline.

**Comment by meltsufin** [comment]
I think the problem occurs due to some auto-configuration classes in spring-cloud-context. In a unit test, we should be able to control which spring-cloud-context auto-configuration is triggered.

**Comment by joaoandremartins** [comment]
@dsyer I see you were the committer of many auto-config classes on `spring-cloud-context`.
Would you have any pointers for why when we include it in a project together with `spring-cloud-gcp-starter`, no `GcpProperties` fields are populated? (But, for instance, `GcpTraceProperties` fields are correctly populated)

I've been trying to reproduce this in a unit test but, unfortunately, I've been unsuccessful so far..

**Comment by meltsufin** [comment]
I think I've narrowed down the problem to 
```
org.springframework.cloud.bootstrap.BootstrapConfiguration=\
org.springframework.cloud.gcp.autoconfigure.config.GcpConfigAutoConfiguration,\
org.springframework.cloud.gcp.autoconfigure.core.GcpContextAutoConfiguration,\
org.springframework.cloud.gcp.autoconfigure.config.GoogleConfigPropertySourceLocator
```
Everything is good once you remove `GcpContextAutoConfiguration` from the list. I think what happens is that during bootstrap, the `GcpProperties` bean is being initialized from `bootstrap.properties`, but it doesn't get re-initialized later for the regular application context from `application.properties`.

**Comment by ChengyuanZhao** [comment]
@meltsufin nice ! I see the fix commit. 

Just out of curiosity, any ideas why it's only an issue if we don't exclude `spring-cloud-context`?

**Comment by meltsufin** [comment]
I think the loading of the bootstrap context is completely skipped if we don't include `spring-cloud-context`.

**Comment by meltsufin** [comment]
Fixed in #368.
